File: lib/src/main/java/com/orientsec/easysocket/client/ClientInitializer.java
package com.orientsec.easysocket.client;

import androidx.annotation.NonNull;

import com.orientsec.easysocket.Address;

import java.util.List;

/**
 * The `ClientInitializer` interface defines the contract for initializing a client.
 * It provides a method to start the initialization process and an inner `Emitter` interface
 * for sending initialization results.
 */
public interface ClientInitializer {

    /**
     * Starts the client initialization process.
     *
     * @param emitter The `Emitter` instance used to send initialization results.
     */
    void start(@NonNull Emitter emitter);

    /**
     * The `Emitter` interface defines methods for sending client initialization events.
     * It is used to notify the success or failure of the initialization process.
     */
    interface Emitter {

        /**
         * Sends an event indicating that the initialization was successful.
         * Notifies that the client is ready for use with the provided address list.
         *
         * @param addressList The list of addresses associated with the client.
         */
        void postSuccess(@NonNull List<Address> addressList);

        /**
         * Sends an event indicating that the initialization has failed.
         * Notifies that an error occurred during the initialization process.
         *
         * @param t The exception describing the reason for the failure.
         */
        void postFailure(@NonNull Throwable t);
    }
}

File: lib/src/commonMain/kotlin/com/orientsec/easysocket/client/ClientInitializer.kt
package com.orientsec.easysocket.client

import com.orientsec.easysocket.Address

/**
 * 客户端初始化器接口，用于在连接前动态获取服务器地址列表。
 *
 * 当服务器地址不固定或需要通过接口动态获取时，可实现此接口。
 * 在 [com.orientsec.easysocket.Options.Builder.addressList] 未设置的情况下，
 * 库会通过此初始化器获取可用地址。
 *
 * 使用示例：
 * ```kotlin
 * class MyClientInitializer : ClientInitializer {
 *     override suspend fun getAddressList(): Result<List<Address>> {
 *         return try {
 *             val addresses = api.getServerAddresses()
 *             Result.success(addresses)
 *         } catch (e: Exception) {
 *             Result.failure(e)
 *         }
 *     }
 * }
 * ```
 */
interface ClientInitializer {

    /**
     * 获取服务器地址列表。
     * 此方法在协程中调用，可以执行网络请求等耗时操作。
     *
     * @return 包含地址列表的 [Result]，成功时返回非空列表，失败时返回异常
     */
    suspend fun getAddressList(): Result<List<Address>>

}